// ngx-popperjs/esm2022/lib/directives/ngx-popperjs/ngx-popperjs-loose.directive.mjs
// tslint:disable:no-input-rename
import { Directive, Inject, Input } from "@angular/core";
import { NGX_POPPERJS_DEFAULTS } from "../../models/ngx-popperjs-defaults.model";
import { NgxPopperjsDirective } from "./ngx-popperjs.directive";
import * as i0 from "@angular/core";
export class NgxPopperjsLooseDirective extends NgxPopperjsDirective {
    set popperLoose(newValue) {
        this.content = newValue;
    }
    set popperLoosePlacement(newValue) {
        this.placement = newValue;
    }
    set popperLooseTrigger(newValue) {
        this.showTrigger = newValue;
    }
    constructor(changeDetectorRef, elementRef, vcr, popperDefaults = {}) {
        super(changeDetectorRef, elementRef, vcr, popperDefaults);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.0", ngImport: i0, type: NgxPopperjsLooseDirective, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }, { token: NGX_POPPERJS_DEFAULTS }], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.0.0", type: NgxPopperjsLooseDirective, isStandalone: true, selector: "[popperLoose]", inputs: { popperLoose: "popperLoose", popperLoosePlacement: "popperLoosePlacement", popperLooseTrigger: "popperLooseTrigger" }, exportAs: ["popperLoose"], usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.0", ngImport: i0, type: NgxPopperjsLooseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: "[popperLoose]",
                    exportAs: "popperLoose",
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGX_POPPERJS_DEFAULTS]
                }] }], propDecorators: { popperLoose: [{
                type: Input
            }], popperLoosePlacement: [{
                type: Input
            }], popperLooseTrigger: [{
                type: Input
            }] } });
