// stream-chat-angular/esm2015/lib/channel-list/channel-list-toggle.service.js
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged, filter, first } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * The `ChannelListToggleService` can be used to toggle the channel list.
 *
 * @deprecated This service can only be used with [theming-v1](../concepts/theming-and-css.mdx), if you are using [thmeing-v2](../theming/introduction.mdx) please refer to our [responsive layout guide](../code-examples/responsive-layout.mdx)
 */
export class ChannelListToggleService {
    constructor() {
        this.isOpenSubject = new BehaviorSubject(false);
        this.isOpen$ = this.isOpenSubject
            .asObservable()
            .pipe(distinctUntilChanged());
        this.isOpen$.pipe(filter((s) => s)).subscribe(() => {
            this.watchForOutsideClicks();
        });
    }
    /**
     * Opens the channel list.
     */
    open() {
        this.isOpenSubject.next(true);
    }
    /**
     * Closes the channel list.
     */
    close() {
        this.isOpenSubject.next(false);
    }
    /**
     * Opens the channel list if it was closed, and closes if it was opened.
     */
    toggle() {
        this.isOpenSubject.getValue() ? this.close() : this.open();
    }
    /**
     * Sets the channel list element, on mobile screen size if the user opens the channel list, and clicks outside, the service automatically closes the channel list if a reference to the HTML element is provided.
     * @param element
     */
    setMenuElement(element) {
        this.menuElement = element;
    }
    /**
     * This method should be called if a channel was selected, if on mobile, the channel list will be closed.
     */
    channelSelected() {
        this.close();
    }
    watchForOutsideClicks() {
        if (!this.menuElement) {
            return;
        }
        const eventHandler = (event) => {
            if (!this.menuElement.contains(event.target)) {
                this.close();
                window.removeEventListener('click', eventHandler);
            }
        };
        window.addEventListener('click', eventHandler);
        this.isOpen$
            .pipe(filter((s) => !s), first())
            .subscribe(() => window.removeEventListener('click', eventHandler));
    }
}
ChannelListToggleService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ChannelListToggleService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ChannelListToggleService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ChannelListToggleService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ChannelListToggleService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return []; } });
