// stream-chat-angular/esm2015/lib/message-list/image-load.service.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * The `ImageLoadService` is used to position the scrollbar in the message list
 * @deprecated - This class is no longer used by SDK components as image sizes are fixed
 */
export class ImageLoadService {
    constructor() {
        /**
         * A subject that can be used to notify the message list if an image attachment finished loading
         */
        this.imageLoad$ = new Subject();
    }
}
ImageLoadService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ImageLoadService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ImageLoadService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ImageLoadService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.5", ngImport: i0, type: ImageLoadService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return []; } });
